MoreLinq/Memoize.cs

```diff
@@ -0,0 +1,136 @@
+        /// in-memory source.
+        /// An additional argument specifies if buffering must happen even if the source implements
+        /// <see cref="ICollection{T}"/> (otherwise it is assumed that the source already serves as 
+        /// a buffer adequately).
+        /// </summary>
+        /// <param name="source">The source sequence.</param>
+        /// <param name="forceBuffering">Force buffering, even if source is an <see cref="ICollection{T}" />
+        /// (otherwise source is assumed to already be in-memory and enumerated quickly, and hence that
+        /// buffering again would be a waste).</param>
+        /// <returns>A <see cref="MemoizedEnumerable{T}"/>.</returns>
+        public static IBufferedEnumerable<T> Memoize<T>(this IEnumerable<T> source, bool forceBuffering)
+        {
+            if (source == null) throw new ArgumentNullException("source");
+
+            return (source as MemoizedEnumerable<T>) ?? new MemoizedEnumerable<T>(source, forceBuffering);
+        }
+    }
+
+    /// <summary>
+    /// An enumeration that, when being iterated, lazily iterates through the source enumeration 
+    /// and cache the elements already iterated for future iterations. It supports partial iterations.
+    /// </summary>
+    /// <typeparam name="T">Type of the source sequence</typeparam>
+    public interface IBufferedEnumerable<T> : IEnumerable<T>
+    {
+
+    }
+
+    internal class MemoizedEnumerable<T> : IBufferedEnumerable<T>, IEnumerable<T>
+    {
+        private ICollection<T> collection;
+        private IEnumerable<T> source;
+        private IEnumerator<T> e;
+        private IList<T> cache;
+        private bool disposed;
+
+        public MemoizedEnumerable(IEnumerable<T> sequence, bool forceBuffering)
+        {
+            if (sequence == null) throw new ArgumentNullException("sequence");
+		
+            if (!forceBuffering && sequence is ICollection<T>)
+            {
+                collection = (ICollection<T>)sequence;
+            }
+            else
+            {
+                source = sequence;
+                cache = new List<T>();
+            }
+        }
+
+        private IEnumerator<T> GetMemoizedEnumerator()
+        {
+            if (e == null && !disposed) e = source.GetEnumerator();
+
+            int index = 0;
+
+            while (true)
+            {
+                while (index < cache.Count)
+                {
+                    yield return cache[index++];
+                }
+
+                if (!disposed && e.MoveNext())
+                {
+                    cache.Add(e.Current);
+                    index++;
+                    yield return e.Current;
+                }
+                else
+                {
+                    if (!disposed)
+                    {
+                        disposed = true;
+                        e.Dispose();
+                        source = null;
+                        e = null;
+                    }
```

Comment:
> Maybe this intermediate approach can be useful: MemoizeEnumerable can have a `Dispose` method (so it can be used within a `using` block), but not implement `IDisposable`, so as not to trigger static analysis tools.

`MemoizeEnumerable` is not exposed. `Memoize` returns `IEnumerable<T>`. Can static analysis tools then see through the implementation that it will return the one and only `MemoizeEnumerable` and which will require disposal?

> It is my suspicion that most memoized enumerables will enumerate completely, and thus explicit disposal is not required.

I too believe this will be 80% use case. That is, memoization will be used to avoid reiterating into a sequence, even if the computations are relatively cheap, without first materializing it in full. For example:

```c#
var q =
    from xs in new[]
    {
        "12 23 34 45",
        "56 67 78 89",
        "1 2 3 4 5 6 7 8 9",
    }
    select (
        from s in xs.Split()
        where s.Length > 0
        select int.Parse(s))
    into xs
    select xs.Memoize() into xs
    select new
    {
        Len = xs.Count(),
        Min = xs.Min(),
        Max = xs.Max(),
        Sum = xs.Sum(),
        Avg = xs.Average(),
    };
```

Without `Memoize` or a `ToArray`, the strings would be re-split and parsed to integers for each of the aggregations in the final projection. For the 20% advanced cases where you'll have to be aware of underlying resources (like a file or database source) that will need disposal, there's the possibility to wrap the scope of use in a `using` or [`Using`](https://github.com/Reactive-Extensions/Rx.NET/blob/v3.1.1/Ix.NET/Source/System.Interactive/Using.cs). Perhaps for those we can introduce a more generalized `WithUsing` that goes like this:

```c#
public static TResult WithUsing<T, TResult>(this IEnumerable<T> source, Func<IEnumerable<T>, TResult> selector)
{
    using (source as IDisposable)
        return selector(source);
}
```

Then the previous example becomes:

```c#
var q =
    from xs in new[]
    {
        "12 23 34 45",
        "56 67 78 89",
        "1 2 3 4 5 6 7 8 9",
    }
    select (
        from s in xs.Split()
        where s.Length > 0
        select int.Parse(s)) into xs
    select xs.Memoize().WithUsing(m => new
    {
        Len = m.Count(),
        Min = m.Min(),
        Max = m.Max(),
        Sum = m.Sum(),
        Avg = m.Average(),
    });
```

How's your chaining looking now? :smile: